Model/Datasource/LdapSource.php

```diff
@@ -230,7 +230,7 @@ public function connect($bindDN = null, $passwd = null) {
 		if ($config['tls']) {
 			if (!ldap_start_tls($this->database)) {
 				$this->log("Ldap_start_tls failed", 'ldap.error');
-				fatal_error("Ldap_start_tls failed");
+				return $this->disconnect();
```

Comment:
I would prefer throwing Exception here over returning false. For example, if it were Mysql connection for example, it would throw PDOException if it failed to connect to the server.

Reply:
Agreed, but nowhere else in this project currently throws an Exception.  There are various places in this class where we could:
https://github.com/cakephp/datasources/blob/84cc8d0b2fd540a2980319ef7594110153eb85cc/Model/Datasource/LdapSource.php#L548-L549
https://github.com/cakephp/datasources/blob/84cc8d0b2fd540a2980319ef7594110153eb85cc/Model/Datasource/LdapSource.php#L597-L598
https://github.com/cakephp/datasources/blob/84cc8d0b2fd540a2980319ef7594110153eb85cc/Model/Datasource/LdapSource.php#L748-L749

Shouldn't refactoring to use Exceptions be a different Issue/PR?

Reply:
Fair point. Now I am working on test cases for LdapSource. After I finished the work, I will merge this PR and #114. But I am not sure if I can merge #115, as it contains some conflicts. If your are still interested in it, then please rebase it and make sure test cases pass.

Reply:
I'm not seeing the merge conflict for #115.  Each of these PRs does have a failed test, but it looked to me like it is all the same failure, outside of the scope of this class:
> 1) XmlrpcSourceTest::testRequest
> Failed asserting that -32300 matches expected -32700.

Can you point me to any checks or conflicts that should be addressed?

Reply:
I have already merged #114 on my local environment to write test cases. And next, I try to merge #115, but then it reported several conflict.

Reply:
Ah, I see.  When you've merged #114 into GitHub, I will be able to rebase #115 to resolve the conflicts.